File: masker/contracts.py
"""Typed contracts shared between Codex (detection/policy), Cursor (integration),
and Ona (UI/trace). These mirror the JSON shapes defined in AGENTS.md so all
three workstreams can build against stable interfaces in parallel.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Literal


class EntityType(str, Enum):
    SSN = "ssn"
    PHONE = "phone"
    EMAIL = "email"
    NAME = "name"
    ADDRESS = "address"
    INSURANCE_ID = "insurance_id"
    MRN = "mrn"
    DOB = "dob"
    HEALTH_CONTEXT = "health_context"
    OTHER = "other"


RiskLevel = Literal["none", "low", "medium", "high"]
Route = Literal["local-only", "masked-send", "safe-to-send"]
PolicyName = Literal[
    "hipaa_base",
    "hipaa_logging_strict",
    "hipaa_clinical_context",
]


@dataclass(frozen=True)
class Entity:
    """A single sensitive span detected in text."""

    type: EntityType
    value: str
    start: int = -1
    end: int = -1
    confidence: float = 1.0
    health_context: bool = False

    def to_dict(self) -> dict[str, Any]:
        return {
            "type": self.type.value,
            "value": self.value,
            "start": self.start,
            "end": self.end,
            "confidence": self.confidence,
            "health_context": self.health_context,
        }


@dataclass(frozen=True)
class DetectionResult:
    """Codex → Cursor / Ona. JSON shape from AGENTS.md:

        {"entities": [{"type": "ssn", "value": "..."}], "risk_level": "high"}
    """

    entities: list[Entity]
    risk_level: RiskLevel
    health_context: bool = False

    def to_dict(self) -> dict[str, Any]:
        return {
            "entities": [e.to_dict() for e in self.entities],
            "risk_level": self.risk_level,
            "health_context": self.health_context,
        }

    @property
    def has_sensitive(self) -> bool:
        return self.risk_level in ("medium", "high")


@dataclass(frozen=True)
class PolicyDecision:
    """Codex → Cursor. JSON shape from AGENTS.md:

        {"route": "masked-send", "policy": "hipaa_base"}
    """

    route: Route
    policy: PolicyName
    reasons: list[str] = field(default_factory=list)
    rationale: str = ""

    def to_dict(self) -> dict[str, Any]:
        return {
            "route": self.route,
            "policy": self.policy,
            "reasons": list(self.reasons),
            "rationale": self.rationale,
        }


@dataclass(frozen=True)
class MaskedText:
    """Codex → Cursor. The user-safe version of the text plus a token map
    so the original values can be restored on the way back out.
    """

    text: str
    token_map: dict[str, str] = field(default_factory=dict)
    replacements: list[dict[str, Any]] = field(default_factory=list)

    def to_dict(self) -> dict[str, Any]:
        return {
            "text": self.text,
            "token_map": dict(self.token_map),
            "replacements": list(self.replacements),
        }


TraceStage = Literal[
    "stt", "detection", "policy", "masking", "routing", "llm", "output_filter", "tts"
]


@dataclass(frozen=True)
class TraceEvent:
    """All → Ona. JSON shape from AGENTS.md:

        {"stage": "masking", "message": "Masked SSN"}
    """

    stage: TraceStage
    message: str
    elapsed_ms: float = 0.0
    payload: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        return {
            "stage": self.stage,
            "message": self.message,
            "elapsed_ms": self.elapsed_ms,
            "payload": dict(self.payload),
        }


@dataclass
class TurnResult:
    """End-to-end output of a single voice turn. Returned by the voice loop
    and consumed by the trace UI / external integrations.
    """

    user_text: str
    detection: DetectionResult
    policy: PolicyDecision
    masked_input: MaskedText
    model_output: str
    safe_output: str
    trace: list[TraceEvent]
    total_ms: float

    def to_dict(self) -> dict[str, Any]:
        return {
            "user_text": self.user_text,
            "detection": self.detection.to_dict(),
            "policy": self.policy.to_dict(),
            "masked_input": self.masked_input.to_dict(),
            "model_output": self.model_output,
            "safe_output": self.safe_output,
            "trace": [t.to_dict() for t in self.trace],
            "total_ms": self.total_ms,
        }

File: platform/masker-core/crates/masker/src/contracts.rs
//! Typed contracts shared between every workstream.
//!
//! These mirror the JSON shapes from `AGENTS.md` so the integration layer,
//! the detection/policy layer, and the trace UI can all build against stable
//! interfaces. The Python reference implementation lives in
//! `platform/masker-library/masker/contracts.py`.

use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EntityType {
    Ssn,
    Phone,
    Email,
    Name,
    Address,
    InsuranceId,
    Mrn,
    Dob,
    HealthContext,
    Other,
    RoutingNumber,
    AccountNumber,
    Pin,
    IpAddress,
}

impl EntityType {
    pub fn as_str(self) -> &'static str {
        match self {
            EntityType::Ssn => "ssn",
            EntityType::Phone => "phone",
            EntityType::Email => "email",
            EntityType::Name => "name",
            EntityType::Address => "address",
            EntityType::InsuranceId => "insurance_id",
            EntityType::Mrn => "mrn",
            EntityType::Dob => "dob",
            EntityType::HealthContext => "health_context",
            EntityType::Other => "other",
            EntityType::RoutingNumber => "routing_number",
            EntityType::AccountNumber => "account_number",
            EntityType::Pin => "pin",
            EntityType::IpAddress => "ip_address",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum RiskLevel {
    None,
    Low,
    Medium,
    High,
}

impl RiskLevel {
    pub fn as_str(self) -> &'static str {
        match self {
            RiskLevel::None => "none",
            RiskLevel::Low => "low",
            RiskLevel::Medium => "medium",
            RiskLevel::High => "high",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum Route {
    LocalOnly,
    MaskedSend,
    SafeToSend,
}

impl Route {
    pub fn as_str(self) -> &'static str {
        match self {
            Route::LocalOnly => "local-only",
            Route::MaskedSend => "masked-send",
            Route::SafeToSend => "safe-to-send",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[allow(clippy::enum_variant_names)] // HIPAA prefix is intentional — policy family
pub enum PolicyName {
    HipaaBase,
    HipaaLogging,
    HipaaClinical,
    GdprBase,
}

impl PolicyName {
    pub fn as_str(self) -> &'static str {
        match self {
            PolicyName::HipaaBase => "hipaa_base",
            PolicyName::HipaaLogging => "hipaa_logging",
            PolicyName::HipaaClinical => "hipaa_clinical",
            PolicyName::GdprBase => "gdpr_base",
        }
    }
}

/// A single sensitive span detected in text.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Entity {
    #[serde(rename = "type")]
    pub kind: EntityType,
    pub value: String,
    /// Byte offsets into the original UTF-8 string.
    pub start: usize,
    pub end: usize,
    pub confidence: f32,
}

impl Entity {
    pub fn new(kind: EntityType, value: impl Into<String>, start: usize, end: usize) -> Self {
        Self {
            kind,
            value: value.into(),
            start,
            end,
            confidence: 0.9,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DetectionResult {
    pub entities: Vec<Entity>,
    pub risk_level: RiskLevel,
}

impl DetectionResult {
    pub fn has_sensitive(&self) -> bool {
        matches!(self.risk_level, RiskLevel::Medium | RiskLevel::High)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PolicyDecision {
    pub route: Route,
    pub policy: PolicyName,
    #[serde(default)]
    pub rationale: String,
}

/// User-safe version of the text plus a token map for re-hydration.
/// Stored as BTreeMap so JSON output is deterministic.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MaskedText {
    pub text: String,
    pub token_map: BTreeMap<String, String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TraceStage {
    Stt,
    Detection,
    Policy,
    Masking,
    Routing,
    Llm,
    OutputFilter,
    Tts,
}

impl TraceStage {
    pub fn as_str(self) -> &'static str {
        match self {
            TraceStage::Stt => "stt",
            TraceStage::Detection => "detection",
            TraceStage::Policy => "policy",
            TraceStage::Masking => "masking",
            TraceStage::Routing => "routing",
            TraceStage::Llm => "llm",
            TraceStage::OutputFilter => "output_filter",
            TraceStage::Tts => "tts",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TraceEvent {
    pub stage: TraceStage,
    pub message: String,
    pub elapsed_ms: f64,
    #[serde(default)]
    pub payload: serde_json::Map<String, serde_json::Value>,
}

/// End-to-end output of a single voice turn.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TurnResult {
    pub user_text: String,
    pub detection: DetectionResult,
    pub policy: PolicyDecision,
    pub masked_input: MaskedText,
    pub model_output: String,
    pub safe_output: String,
    pub trace: Vec<TraceEvent>,
    pub total_ms: f64,
}
